feat(napi/parser)!: add typings to napi/parser

ottomated opened the pull request:
- Added TypeScript annotation for `ParseResult.program`
- Modified the entrypoint for `oxc-parser` to allow wrapping the napi functions
- Updated `index.js` to parse the `program` string into a JSON object
- Updated tests
- Added a dependency on `@oxc/types`

ottomated commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6796?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#6796** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6796?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#6755** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6755?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @ottomated and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

overlookmotel commented:
Note: This PR is a breaking change to `oxc-parser` package, as `parseSync` and `parseAsync` now return the AST as a JS object, rather than JSON string. That change is good (#4295), but we should make sure it's released with suitable semver bump.

overlookmotel commented:
One other thing: NPM package README should be updated to reflect the API change.

https://github.com/oxc-project/oxc/blob/main/npm/oxc-parser/README.md

overlookmotel reviewed (changes requested):
NPM package README should be updated to reflect the API change.

https://github.com/oxc-project/oxc/blob/main/npm/oxc-parser/README.md

(please rebase on top of #6834 once its merged, to avoid merge conflicts)

overlookmotel reviewed (changes requested):
Please update [dprint config](https://github.com/oxc-project/oxc/blob/main/dprint.json) to add `bindings.js` to the exclusion list (and remove `index.js`).

That'll stop autofix CI job repeatedly reformatting NAPI-RS's generated code.

ottomated commented on napi/parser/package.json:
```diff
@@ -2,8 +2,9 @@
+  "type": "module",
   "scripts": {
-    "build": "napi build --platform --release",
+    "build": "napi build --platform --release --js bindings.cjs",
```
The package is now esm because of the index.js wrapper

overlookmotel commented on napi/parser/index.js:
```diff
@@ -1,14 +1,15 @@
-  result.program = JSON.parse(result.program);
-  return result;
-}
+module.exports = {
+  ...bindings,
+  parseSync(...args) {
+    const result = bindings.parseSync(...args);
+    result.program = JSON.parse(result.program);
+    return result;
+  },
+  async parseAsync(...args) {
+    const result = await bindings.parseAsync(...args);
+    result.program = JSON.parse(result.program);
+    return result;
+  },
+};
```
So that NodeJS's CJS module lexer can see all the exports and give a nice result when you `import` the package, would be better to export each individually, like NAPI-RS does:

https://github.com/oxc-project/oxc/blob/2de9ef578ecac4d9c99cfca171a0dbdbb4d9a66c/napi/parser/bindings.js#L345-L349

Sorry to be a pain!

ottomated replied:
Yikes, this is why moving to esm would be nice. This is going to be a pain to maintain.

overlookmotel commented on tasks/ast_tools/src/generators/typescript.rs:
```diff
@@ -27,8 +27,8 @@ impl Generator for TypescriptGenerator {
             "\
-        		// To edit this generated file you have to edit `{file}`\n\
-        		// Auto-generated code, DO NOT EDIT DIRECTLY!\n\n\
+        		// Auto-generated code, DO NOT EDIT DIRECTLY!\n\
+        		// To edit this generated file you have to edit `{file}`\n\n\
```
Indentation still looks weird here. Are there still tabs? If your IDE isn't letting you use spaces for some reason, no worries, I can fix in a follow-up.

overlookmotel reviewed (changes requested):
This should be the last round of change requests!

ottomated commented on npm/oxc-parser/.gitignore:
```diff
@@ -0,0 +1,4 @@
+node_modules/
+index.js
+bindings.js
+index.d.ts
```
They're in the napi/parser directory, oxc-parser's generate-packages.mjs script copies them into here where they aren't checked into git.

overlookmotel replied:
Ah ha. Now I understand. Thanks for explaining.

overlookmotel reviewed (approved):
Great! All we need to do is resolve question of what package name to publish the types under, and we're good to go.

overlookmotel commented:
Note: Merge #6836 too after merging this.

Boshen commented:
### Merge activity

* **Oct 24, 9:07 AM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

ottomated committed:
feat(napi/parser)!: add typings to napi/parser (#6796)

- Added TypeScript annotation for `ParseResult.program`
- Modified the entrypoint for `oxc-parser` to allow wrapping the napi functions
- Updated `index.js` to parse the `program` string into a JSON object
- Updated tests
- Added a dependency on `@oxc/types`

Dunqing commented:
We don't have `oxc` scope in NPM if I remember correctly @Boshen cc
